src/utils/object.js

```diff
@@ -44,3 +46,63 @@ export function map(opt_initial) {
 export function hasOwn(obj, key) {
   return hasOwn_.call(obj, key);
 }
+
+/**
+ * @param {!Object} target
+ * @param {!Object} source
+ * @param {number} maxDepth The maximum depth for deep merge, beyond which
+ *    Object.assign will be used.
+ * @return {!Object}
+ * @throws {Error} if `source` contains a circular reference
+ */
+function deepMerge_(target, source, maxDepth) {
+  // Keep track of seen objects to prevent infinite loops on objects with
+  // recursive references.
+  const seen = [];
+  // Traversal must be breadth-first so any object encountered for the first
+  // time does not have a reference  at a shallower depth. Otherwise, a
+  // circular reference found at depth == maxDepth could cause an unexpected
+  // change at a shallower depth if the same object exists at a shallower depth.
+  const queue = [{target, source, depth: 0}];
+  while (queue.length > 0) {
+    const {target, source, depth} = queue.shift();
+    if (seen.includes(source)) {
+      // No recursive merge
+      throw new Error('deepMerge: Source object contains circular references');
```

Comment:
Nit: The `deepMerge:` prefix is unnecessary since the stack will show its origin.